src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -670,18 +689,78 @@ private Compilation CreateEmptyCompilation()
                 return finalState.HasSuccessfullyLoaded;
             }
 
-            public ICompilationTracker FreezePartialState(CancellationToken cancellationToken)
+            public ICompilationTracker WithCreationPolicy(bool create, bool forceRegeneration, CancellationToken cancellationToken)
+            {
+                return create
+                    ? WithCreateCreationPolicy(forceRegeneration)
+                    : WithDoNotCreateCreationPolicy(forceRegeneration, cancellationToken);
+            }
+
+            public ICompilationTracker WithCreateCreationPolicy(bool forceRegeneration)
+            {
+                var state = this.ReadState();
+
+                var desiredCreationPolicy = CreationPolicy.Create;
+
+                // If we've computed no state yet there's nothing to do.  This state will automatically transition 
+                // to an InProgressState with a creation policy of 'Create' anyways.
+                if (state is null)
+                    return this;
+
+                // If we're already in the state where we are running generators and skeletons we don't need to do
+                // anything and can just return ourselves. The next request to create the compilation will do so fully,
+                // including running generators.
+                if (state.CreationPolicy == desiredCreationPolicy)
+                    return this;
+
+                InProgressState newState;
+                if (state is InProgressState inProgressState)
+                {
+                    newState = new InProgressState(
+                        desiredCreationPolicy,
+                        inProgressState.LazyCompilationWithoutGeneratedDocuments,
+                        forceRegeneration ? inProgressState.GeneratorInfo with { Driver = null } : inProgressState.GeneratorInfo,
```

Comment:
if we're forcing regeneration, then we drop the generator driver.  this means we will always rerun the generator.